File: Algorithms_Data_Structures/Inventory_Management_System/Inventory.java
import java.util.HashMap;

public class Inventory {
    private HashMap<String, Product> productMap;

    public Inventory() {
        productMap = new HashMap<>();
    }

    public void addProduct(Product product) {
        if (productMap.containsKey(product.getProductId())) {
            System.out.println("Product ID already exists. Use update instead.");
            return;
        }
        productMap.put(product.getProductId(), product);
        System.out.println("Added: " + product.getProductName());
    }

    public void updateProduct(String productId, int newQuantity, double newPrice) {
        Product product = productMap.get(productId);
        if (product != null) {
            product.setQuantity(newQuantity);
            product.setPrice(newPrice);
            System.out.println("Updated ID " + productId + " successfully.");
        } else {
            System.out.println("Product not found to update.");
        }
    }

    public void deleteProduct(String productId) {
        if (productMap.containsKey(productId)) {
            Product removed = productMap.remove(productId);
            System.out.println("Deleted: " + removed.getProductName());
        } else {
            System.out.println("Product not found to delete.");
        }
    }

    public void displayInventory() {
        System.out.println("\n--- Current Inventory Status ---");
        for (Product p : productMap.values()) {
            System.out.println(p);
        }
    }
}

File: Algorithms_Data_Structures/Inventory_Management_System/Inventory.cs
using System;
using System.Collections.Generic;

public class Inventory {
    // Dictionary is C#'s version of a HashMap
    private Dictionary<string, Product> productMap;

    public Inventory() {
        productMap = new Dictionary<string, Product>();
    }

    public void AddProduct(Product product) {
        if (productMap.ContainsKey(product.GetProductId())) {
            Console.WriteLine("Product ID already exists. Use update instead.");
            return;
        }
        productMap.Add(product.GetProductId(), product);
        Console.WriteLine("Added: " + product.GetProductName());
    }

    public void UpdateProduct(string productId, int newQuantity, double newPrice) {
        if (productMap.TryGetValue(productId, out Product product)) {
            product.SetQuantity(newQuantity);
            product.SetPrice(newPrice);
            Console.WriteLine("Updated ID " + productId + " successfully.");
        } else {
            Console.WriteLine("Product not found to update.");
        }
    }

    public void DeleteProduct(string productId) {
        if (productMap.ContainsKey(productId)) {
            Product removed = productMap[productId];
            productMap.Remove(productId);
            Console.WriteLine("Deleted: " + removed.GetProductName());
        } else {
            Console.WriteLine("Product not found to delete.");
        }
    }

    public void DisplayInventory() {
        Console.WriteLine("\n--- Current Inventory Status ---");
        foreach (Product p in productMap.Values) {
            Console.WriteLine(p);
        }
    }
}